subprojects/core/src/main/java/org/gradle/api/tasks/AbstractCopyTask.java

```diff
@@ -88,9 +92,9 @@ protected AbstractCopyTask() {
             getInputs().property(specPropertyName + ".caseSensitive", (Callable<Boolean>) spec::isCaseSensitive);
             getInputs().property(specPropertyName + ".includeEmptyDirs", (Callable<Boolean>) spec::getIncludeEmptyDirs);
             getInputs().property(specPropertyName + ".duplicatesStrategy", (Callable<DuplicatesStrategy>) spec::getDuplicatesStrategy);
-            getInputs().property(specPropertyName + ".dirMode", (Callable<Integer>) spec::getDirMode)
+            getInputs().property(specPropertyName + ".dirPermissions", spec.getDirPermissions().flatMap(it -> Providers.of(it.toUnixNumeric())))
                 .optional(true);
-            getInputs().property(specPropertyName + ".fileMode", (Callable<Integer>) spec::getFileMode)
+            getInputs().property(specPropertyName + ".filePermissions", spec.getFilePermissions().flatMap(it -> Providers.of(it.toUnixNumeric())))
```

Comment:
:thought_balloon: I think there's an unnecessary `Provider` creation here
```suggestion
            getInputs().property(specPropertyName + ".filePermissions", spec.getFilePermissions().map(ImmutableFileAccessPermissions::toUnixNumeric))
```
This change requires changing imports